src/librustc_session/session.rs

```diff
@@ -55,6 +57,52 @@ pub enum CtfeBacktrace {
     Immediate,
 }
 
+/// New-type wrapper around `usize` for representing limits. Ensures that comparisons against
+/// limits are consistent throughout the compiler.
+#[derive(Clone, Copy, Debug)]
+pub struct Limit(usize);
+
+impl Limit {
+    /// Create a new limit from a `usize`.
+    pub fn new(value: usize) -> Self {
+        Limit(value)
+    }
+
+    /// Check that `value` is within the limit. Ensures that the same comparisons are used
+    /// throughout the compiler, as mismatches can cause ICEs, see #72540.
+    pub fn within_limit(&self, value: usize) -> bool {
```

Comment:
Maybe `value_within_limit` would be a less ambiguous name, because currently it reads like the argument is the limit, rather than the other way around.

Reply:
Bikeshedding: `is_reached_by` so the code reads `recursion_limit.is_reached_by(some_depth)`

Reply:
I've changed the name to this.